fortran/src/H5Dff.F90

```diff
@@ -386,29 +595,149 @@ END SUBROUTINE h5dget_type_f
+         INTEGER(HID_T), VALUE :: dataset_id
+         INTEGER(HSIZE_T), DIMENSION(*) :: size
+       END FUNCTION H5Dset_extent
     END INTERFACE
 
-    hdferr = H5Dset_extent_c(dataset_id, size)
+    CALL H5Dget_space_f(dataset_id, space_id, hdferr)
+    IF(hdferr.LT.0) RETURN
+
+    CALL H5Sget_simple_extent_ndims_f(space_id, rank, hdferr)
+    IF( hdferr.LT.0 .OR. rank.LT.0 )THEN
+       CALL H5Sclose_f(space_id, hdferr)
+       hdferr = -1
+       RETURN
+    ENDIF
+    CALL H5Sclose_f(space_id, hdferr)
+    IF(hdferr.LT.0) RETURN
+
+    ALLOCATE(csize(rank), STAT=hdferr)
+    IF (hdferr .NE. 0 ) THEN
+       hdferr = -1
+       RETURN
+    ENDIF
+
+    !
+    ! Reverse dimensions due to C-FORTRAN storage order.
+    !
+    DO i = 1, rank
+       csize(i) = fsize(rank - i + 1)
+    ENDDO
+
+    hdferr = INT(H5Dset_extent(dataset_id, csize))
+
+    DEALLOCATE(csize)
+
   END SUBROUTINE h5dset_extent_f
 
+
+!>
+!! \ingroup FH5D
+!!
+!! \brief Asynchronously  extends a dataset with unlimited dimension.
+!!
+!! \param dataset_id Dataset identifier
+!! \param fsize      Array containing the new magnitude of each dimension
+!! \param es_id      \es_id
+!! \param hdferr     \fortran_error
+!! \param file       \fortran_file
+!! \param func       \fortran_func
+!! \param line       \fortran_line
+!!
+!! See C API: @ref H5Dset_extent_async()
+!!
+  SUBROUTINE h5dset_extent_async_f(dataset_id, fsize, es_id, hdferr, file, func, line)
+    IMPLICIT NONE
+    INTEGER(HID_T), INTENT(IN) :: dataset_id
+    INTEGER(HSIZE_T), DIMENSION(*), INTENT(IN) :: fsize
+    INTEGER(HID_T), INTENT(IN)  :: es_id
+    INTEGER, INTENT(OUT) :: hdferr
+    CHARACTER(LEN=*), INTENT(IN), OPTIONAL :: file
+    CHARACTER(LEN=*), INTENT(IN), OPTIONAL :: func
+    INTEGER         , INTENT(IN), OPTIONAL :: line
+
+    CHARACTER(LEN=CHR_MAX,KIND=C_CHAR) ::  file_default = C_NULL_CHAR
+    CHARACTER(LEN=CHR_MAX,KIND=C_CHAR) ::  func_default = C_NULL_CHAR
+    INTEGER(KIND=C_INT) :: line_default = 0
+    INTEGER(HSIZE_T), DIMENSION(:), ALLOCATABLE :: csize
+    INTEGER(HID_T) :: space_id
+    INTEGER :: rank
+    INTEGER :: i
+
+    INTERFACE
+       INTEGER(C_INT) FUNCTION H5Dset_extent_async(file, func, line, dataset_id, size, es_id) &
+            BIND(C,NAME='H5Dset_extent_async')
+         IMPORT :: C_CHAR, C_INT
+         IMPORT :: HID_T, HSIZE_T
+         IMPLICIT NONE
+         CHARACTER(KIND=C_CHAR), DIMENSION(*) :: file
+         CHARACTER(KIND=C_CHAR), DIMENSION(*) :: func
+         INTEGER(C_INT), VALUE :: line
+         INTEGER(HID_T), VALUE :: dataset_id
+         INTEGER(HSIZE_T), DIMENSION(*) :: size
+         INTEGER(HID_T), VALUE :: es_id
+       END FUNCTION H5Dset_extent_async
+    END INTERFACE
+
+    IF(PRESENT(file)) file_default = TRIM(file)//C_NULL_CHAR
+    IF(PRESENT(func)) func_default = TRIM(func)//C_NULL_CHAR
+    IF(PRESENT(line)) line_default = INT(line, C_INT)
+
+    CALL H5Dget_space_async_f(dataset_id, space_id, es_id, hdferr)
```

Comment:
Should it be just H5Dget_space_f? If not, do we need to check if the subroutine returns before calling H5Sget_simple_extent_ndims_f?

Reply:
Good catch, it should be a sync version, fixed.